gnovm/pkg/doctest/README.md

```diff
@@ -0,0 +1,131 @@
+# Gno Doctest: Easy Code Execution and Testing
+
+Gno Doctest is a tool that allows you to easily execute and test code blocks written in the Gno language. This tool offers a range of features, from simple code execution to complex package imports.
+
+## Basic Usage
+
+To use Gno Doctest, run the following command:
+
+gno doctest -path <markdown_file_path> -run <code_block_name | "">
+
+- `<markdown_file_path>`: Path to the markdown file containing Gno code blocks
+- `<code_block_name>`: Name of the code block to run (optional)
+
+For example, to run the code block named "print hello world" in the file "foo.md", use the following command:
+
+gno doctest -path foo.md -run "print hello world"
+
+## Features
+
+### 1. Basic Code Execution
+
+Gno Doctest can execute simple code blocks:
+
+```go
+package main
+
+func main() {
+    println("Hello, World!")
+}
+
+// Output:
+// Hello, World!
+```
+
+Doctest also recognizes that a block of code is a gno. The code below outputs the same result as the example above.
+
+```go
```

Comment:
How does Doctest detect this is a gno? because //@test: print hello world?

What is the reason behind using "```go" instead of  "````gno"  for gno code block?

Reply:
That's a really good question. Currrently, when rendering markdown, gno doesn't have universal syntax highlighting applied yet (except the custom syntax rule). So, even in official documents like "Effective gno" are using `go` instead of `gno` to specify the language.

Considering the purpose of markdown is more for documentation rather than testing, this was an inevitable purpose. So I made it recognize both languages.

However, it doesn't clearly distinguish whether it's completely `go` or `gno`. For now, I assume that it's mostly compatible with go and execute it at gno VM level. In other words, even if the language is atually written in go, everything runs as gno. This part may require deeper context analysis in the future.

To summarize in the order of your questions:
  1. Even if the code block is specified as Go, it's actually recognized and executed as gno.
  2. There's no feature to clearly distinguish between the two languages, and it relies on the language specified in the fenced code block.
  3. Go is also set to be recognized due to syntax highlighting.

Reply:
The main difference between Gno and Go is that Gno has a few native methods unique to blockchain usage. For example:

[std.AssertOriginCall](https://github.com/gnolang/gno/blob/912a5dbf5c1d7118472a4f46b26bfcd7f4072856/gnovm/stdlibs/generated.go#L422)
[std.GetHeight](https://github.com/gnolang/gno/blob/912a5dbf5c1d7118472a4f46b26bfcd7f4072856/gnovm/stdlibs/generated.go#L474C4-L474C13)

Basically, a program can either be Go code or Gno code. We cannot mix them in one source file.

There are a few options I can think of

- Users need to explicitly mark it as either Go code or Gno code, similar to indicating the source code type with a suffix in the filename: xxx.go vs xxx.gno.
- Only support Gno.
- By default, the MD is treated as Gno code. Use gnostd.XXX to indicate that it is Go code. (Not sure if this has much practical purpose; it is more complicated to implement and is only for the sake of supporting Go code with chain functionality in MD.)